contracts/cw20-merkle-airdrop/src/contract.rs

```diff
@@ -0,0 +1,486 @@
+    if info.sender != cfg.owner {
+        return Err(ContractError::Unauthorized {});
+    }
+
+    // validate owner change
+    let new_owner = owner.ok_or(ContractError::InvalidInput {})?;
+    let new_owner = deps.api.addr_validate(&new_owner)?;
+
+    CONFIG.update(deps.storage, |mut exists| -> StdResult<_> {
+        exists.owner = new_owner;
+        Ok(exists)
+    })?;
+
+    Ok(Response {
+        messages: vec![],
+        attributes: vec![attr("action", "update_config")],
+        data: None,
+        events: vec![],
+    })
+}
+
+pub fn execute_register_merkle_root(
+    deps: DepsMut,
+    _env: Env,
+    info: MessageInfo,
+    merkle_root: String,
+) -> Result<Response, ContractError> {
+    // authorize owner
+    let cfg = CONFIG.load(deps.storage)?;
+    if info.sender != cfg.owner {
+        return Err(ContractError::Unauthorized {});
+    }
+
+    // check merkle root length
+    let mut root_buf: [u8; 32] = [0; 32];
+    hex::decode_to_slice(merkle_root.to_string(), &mut root_buf)?;
+
+    let stage = STAGE.update(deps.storage, |stage| Ok(stage + 1))?;
+
+    MERKLE_ROOT.save(deps.storage, U8Key::from(stage), &merkle_root)?;
+    STAGE.save(deps.storage, &stage)?;
+
+    Ok(Response {
+        messages: vec![],
+        attributes: vec![
+            attr("action", "register_merkle_root"),
+            attr("stage", stage),
+            attr("merkle_root", merkle_root),
+        ],
+        data: None,
+        events: vec![],
+    })
+}
+
+pub fn execute_claim(
+    deps: DepsMut,
+    _env: Env,
+    info: MessageInfo,
+    stage: u8,
+    amount: Uint128,
+    proof: Vec<String>,
+) -> Result<Response, ContractError> {
+    // verify not claimed
+    let claimed = CLAIM.may_load(deps.storage, (&info.sender, U8Key::from(stage)))?;
+    if claimed.is_some() {
+        return Err(ContractError::Claimed {});
+    }
+
+    let config = CONFIG.load(deps.storage)?;
+    let merkle_root = MERKLE_ROOT.load(deps.storage, stage.into())?;
+
+    let user_input = format!("{}{}", info.sender, amount);
+    let hash = sha3::Keccak256::digest(user_input.as_bytes())
+        .as_slice()
+        .try_into()
+        .map_err(|_| ContractError::WrongLength {})?;
+
+    let hash = proof.into_iter().try_fold(hash, |hash, p| {
+        let mut proof_buf = [0; 32];
+        hex::decode_to_slice(p, &mut proof_buf)?;
+        let mut hashes = [hash, proof_buf];
+        hashes.sort_unstable();
+        sha3::Keccak256::digest(&hashes.concat())
+            .as_slice()
+            .try_into()
+            .map_err(|_| ContractError::WrongLength {})
+    })?;
+
+    let mut root_buf: [u8; 32] = [0; 32];
+    hex::decode_to_slice(merkle_root, &mut root_buf)?;
+    if root_buf != hash {
+        return Err(ContractError::VerificationFailed {});
+    }
+
+    // Update claim index to the current stage
+    CLAIM.save(deps.storage, (&info.sender, stage.into()), &true)?;
+
+    let msgs: Vec<SubMsg> = vec![SubMsg::new(WasmMsg::Execute {
```

Comment:
Note that the contract has no idea if there is enough funds for this.
I guess the funder must do ensure the funds are there before they set the merkle root, right?
This is fine and seems to be documented. Just noticing it.